src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -24075,6 +24075,151 @@ bool Init()
                 );
         }
 
+        [Fact]
+        public void MemberNotNullWhenTrue_EnforcedInMethodBody_NonConstant()
+        {
+            var c = CreateNullableCompilation(new[] { @"
+using System.Diagnostics.CodeAnalysis;
+public class C
+{
+    public string field1;
+    public string? field2;
+    public string field3;
+    public string? field4;
+
+    C() // 1
+    {
+        if (!Init()) throw null!;
+    }
+
+    [MemberNotNullWhen(true, nameof(field1), nameof(field2))]
+    bool Init()
+    {
+        bool b = true;
+        if (b)
+        {
+            return b;
+        }
+        if (b)
+        {
+            return !b;
+        }
+        return M(out _);
+    }
+
+    bool M([MaybeNullWhen(true)]out string s) => throw null!;
+}
+", MemberNotNullWhenAttributeDefinition, MaybeNullWhenAttributeDefinition });
+
+            c.VerifyDiagnostics(
+                // (10,5): warning CS8618: Non-nullable field 'field3' is uninitialized. Consider declaring the field as nullable.
+                //     C() // 1
+                Diagnostic(ErrorCode.WRN_UninitializedNonNullableField, "C").WithArguments("field", "field3").WithLocation(10, 5)
+                );
+        }
+
+        [Fact]
+        public void MemberNotNullWhenTrue_EnforcedInMethodBody_NonConstant_ButAnalyzable()
+        {
+            var c = CreateNullableCompilation(new[] { @"
+using System.Diagnostics.CodeAnalysis;
+public class C
+{
+    public string field1;
+    public string field3;
+
+    C()
+    {
+        if (!Init()) throw null!;
+    }
+
+    [MemberNotNullWhen(true, nameof(field1), nameof(field3))]
+    bool Init()
+    {
+        bool b = true;
+        if (b)
+        {
+            return field1 == null; // 1
```

Comment:
this is an interesting place to land. We give a warning if you perform a check that "contradicts" the nullability attributes, but we don't warn if you neglect to include a check that would affirm the nullability attributes. I think that's OK.

Reply:
I'd say we don't warn if you return something we couldn't understand.

---
In reply to: [502707893](https://github.com/dotnet/roslyn/pull/48425#discussion_r502707893) [](ancestors = 502707893)